cmd/fluxd/config.go

```diff
@@ -0,0 +1,163 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+	"time"
+
+	"github.com/spf13/pflag"
+	"github.com/spf13/viper"
+
+	"github.com/fluxcd/flux/pkg/config"
+	"github.com/fluxcd/flux/pkg/registry"
+	"github.com/fluxcd/flux/pkg/sync"
+)
+
+// defineConfigFlags defines the flags that can also be set in
+// a config file. These need special treatment, because some care must
+// be taken to match them ("bind") with config file field names.
+func defineConfigFlags(fs *pflag.FlagSet, bail func(error)) {
+
+	bind := func(fieldName, flagName string) error {
+		configStruct := reflect.TypeOf(config.Config{})
+		field, ok := configStruct.FieldByName(fieldName)
+		if !ok {
+			return fmt.Errorf("attempt to bind a flag to a field not present in config.Config, %q", fieldName)
+		}
+		tag := field.Tag
+		// this parallels the logic in
+		// github.com/mitchellh/mapstructure, except that we want to
+		// bail if a field is mentioned that is marked ignore, like
+		// this: `mapstructure:"-"`
+		mappedName := field.Name
+		mapstructureTagParts := strings.Split(tag.Get("mapstructure"), ",")
+		if namePart := mapstructureTagParts[0]; namePart != "" {
+			if namePart == "-" { // means ignore this field
+				return fmt.Errorf(`attempt to bind a flag to a config field tagged as ignored, %q`, field.Name)
+			}
+			mappedName = namePart
+		}
+		return viper.BindPFlag(mappedName, fs.Lookup(flagName))
+	}
+
+	bindOrBail := func(flagName, fieldName string) {
+		if err := bind(flagName, fieldName); err != nil {
+			bail(err)
+		}
+	}
+
+	defineString := func(fieldName, flagName, def, desc string) {
+		fs.String(flagName, def, desc)
+		bindOrBail(fieldName, flagName)
+	}
+
+	defineStringP := func(fieldName, flagName, short, def, desc string) {
+		fs.StringP(flagName, short, def, desc)
+		bindOrBail(fieldName, flagName)
+	}
+
+	defineStringSlice := func(fieldName, flagName string, def []string, desc string) {
+		fs.StringSlice(flagName, def, desc)
+		bindOrBail(fieldName, flagName)
+	}
+
+	defineBool := func(fieldName, flagName string, def bool, desc string) {
+		fs.Bool(flagName, def, desc)
+		bindOrBail(fieldName, flagName)
+	}
+
+	defineDuration := func(fieldName, flagName string, def time.Duration, desc string) {
+		fs.Duration(flagName, def, desc)
+		bindOrBail(fieldName, flagName)
+	}
+
+	defineInt := func(fieldName, flagName string, def int, desc string) {
+		fs.Int(flagName, def, desc)
+		bindOrBail(fieldName, flagName)
+	}
+
+	defineFloat64 := func(fieldName, flagName string, def float64, desc string) {
+		fs.Float64(flagName, def, desc)
+		bindOrBail(fieldName, flagName)
+	}
+
+	defineString("LogFormat", "log-format", "fmt", "change the log format.")
```

Comment:
I was hoping for the `define` functions to automate the transforming from flag format into config-file option format (e.g. from `log-format` to `LogFormat`)

Reply:
Ah, I see that acronyms make it a bit hard (e.g. `ci` to `CI`) . But `CI` to `ci` is straight-forward. So maybe we can do it the other way around.

Reply:
The flags are not always straight-forward transliterations of the field names. For example: `k8s-unsafe-exclude-resource` vs `K8sExcludeResource`; `git-ci-skip` vs `GitCISkip`. Could it be made to line up? Probably, since it's the keys (that appear in the mapstructure tag) that we care about, less so the field names, which wouldn't pass linting as they are anyway.

Reply:
We don't need to fix it right now.

However, the acronyms case can be dealt with by counting consecutive capital letters.

As for mismatches like `Unsafe` there are alternatives (e.g. using `K8sUnsafeExcludeResource`, which frankly I think we should be using or only providing names for exceptions).

One neat way to solve this would be `reflect`ing on the `Config` struct (we could use annotations for the name exceptions). We could even include the flag metadata in the struct itself and make the `bind` calls automatically.